contracts/puppeteer-native/src/contract.rs

```diff
@@ -0,0 +1,531 @@
+        msg.allowed_senders.as_ref(),
+        None,
+    )?;
+    let owner = deps
+        .api
+        .addr_validate(&msg.owner.unwrap_or(info.sender.to_string()))?
+        .to_string();
+
+    let config = &Config {
+        remote_denom: msg.remote_denom,
+        allowed_senders: allowed_senders.clone(),
+        native_bond_provider: deps.api.addr_validate(&msg.native_bond_provider)?,
+    };
+
+    let attrs: Vec<Attribute> = vec![
+        attr("owner", &owner),
+        attr("remote_denom", &config.remote_denom),
+        attr("native_bond_provider", &config.native_bond_provider),
+        attr(
+            "allowed_senders",
+            allowed_senders
+                .into_iter()
+                .map(|addr| addr.to_string())
+                .collect::<Vec<_>>()
+                .join(","),
+        ),
+    ];
+
+    cw_ownable::initialize_owner(deps.storage, deps.api, Some(&owner))?;
+    CONFIG.save(deps.storage, config)?;
+
+    Ok(response("instantiate", CONTRACT_NAME, attrs))
+}
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn query(deps: Deps<NeutronQuery>, env: Env, msg: QueryMsg) -> ContractResult<Binary> {
+    match msg {
+        QueryMsg::Extension { msg } => match msg {
+            QueryExtMsg::Delegations {} => query_delegations(deps, env),
+            QueryExtMsg::Balances {} => query_balances(deps, env),
+            // QueryExtMsg::NonNativeRewardsBalances {} => {
+            //     query_non_native_rewards_balances(deps, env)
+            // }
+            QueryExtMsg::UnbondingDelegations {} => query_unbonding_delegations(deps, env),
+            QueryExtMsg::Ownership {} => {
+                let owner = cw_ownable::get_ownership(deps.storage)?;
+                to_json_binary(&owner).map_err(ContractError::Std)
+            }
+        },
+        QueryMsg::Config {} => query_config(deps),
+        QueryMsg::Transactions {} => query_transactions(deps),
+    }
+}
+
+fn query_config(deps: Deps<NeutronQuery>) -> ContractResult<Binary> {
+    let config = CONFIG.load(deps.storage)?;
+    Ok(to_json_binary(&config)?)
+}
+
+fn query_transactions(deps: Deps<NeutronQuery>) -> ContractResult<Binary> {
+    let transfers: Vec<Transfer> = RECIPIENT_TRANSFERS.load(deps.storage)?;
+    Ok(to_json_binary(&transfers)?)
+}
+
+fn query_delegations(deps: Deps<NeutronQuery>, env: Env) -> ContractResult<Binary> {
+    let mut key = vec![];
+    let mut total_delegations = vec![];
+
+    loop {
+        let res: StdResult<QueryDelegatorDelegationsResponse> =
+            deps.querier.query(&QueryRequest::Stargate {
+                path: "/cosmos.staking.v1beta1.Query/DelegatorDelegations".to_string(),
+                data:
+                    cosmos_sdk_proto::cosmos::staking::v1beta1::QueryDelegatorDelegationsRequest {
+                        delegator_addr: env.contract.address.to_string(),
+                        pagination: Some(PageRequest {
+                            key: key.clone(),
```

Comment:
```suggestion
                            key,
```

This clone is absolutely useless here.